### PR TITLE
Initialize from code string

### PR DESCRIPTION
**Pull request by Epigene**
I could use the ability to initialize not from a file but from raw ruby code string.

### CODE REVIEW

**Comment by Epigene** [comment]
I am getting some failures in Ryan::Const#name specs.
Gist: https://gist.github.com/Epigene/e56d433e6928c4612df1af43505d056d

File: lib/ryan.rb [hunk 1]
```diff
@@ -18,9 +18,25 @@ class Ryan
   def_delegators :const,
     :name, :funcs, :type, :initialization_args, :func_by_name, :class?, :module?
 
+  def self.root
```
Comment (ridiculous) [inline lib/ryan.rb]: Good suggestion! I like the idea of having a `root` provided by gems, especially if they deal a lot with files!

File: lib/ryan.rb [hunk 2]
```diff
@@ -18,9 +18,25 @@ class Ryan
...
+    current_file_name = __FILE__
+    path_to_current_file = File.expand_path(current_file_name)
+    root_path = Pathname.new(path_to_current_file) + "../.."
+
+    return root_path
```
Comment (ridiculous) [inline lib/ryan.rb]: More conventional and concise to simply do `Pathname.new File.expand_path('../..', __FILE__)`. You have all the right the pieces there, just organized a bit differently.

File: lib/ryan.rb [hunk 3]
```diff
@@ -18,9 +18,25 @@ class Ryan
...
+  end
+
   # @param [Pathname, String] file
-  def initialize(file)
-    @sexp = RubyParser.new.parse File.read(file)
+  def initialize(file, mode=:file)
```
Comment (ridiculous) [inline lib/ryan.rb]: I'd prefer we kept the same interface, but make it more flexible. How bout we delegate to `File`?
```ruby
input = File.read(input) if File.file?(input)
@sexp = RubyParser.new.parse input
```

File: ryan.gemspec [hunk 4]
```diff
@@ -18,10 +18,13 @@ Gem::Specification.new do |spec|
...
 
+  spec.required_ruby_version = '>= 1.9.3'
+
   spec.add_development_dependency "bundler", "~> 1.10"
   spec.add_development_dependency "rake", "~> 10.0"
   spec.add_development_dependency "rspec", ">= 3.2", "< 4"
 
   spec.add_dependency 'ruby_parser', '>= 3.7.0', '< 4.0.0'
   spec.add_dependency 'ruby2ruby', '~> 2.2'
+  spec.add_dependency 'pry', '~> 0.10.4'
```
Comment (ridiculous) [inline ryan.gemspec]: Let's move this to a development dependency

File: spec/lib/ryan.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
+require "spec_helper"
+# rspec spec/lib/ryan.rb
+describe Ryan do
+  klass = Ryan
```
Comment (ridiculous) [inline spec/lib/ryan.rb]: Can use `described_class` here instead (provided by RSpec)

**Review by ridiculous — COMMENTED** [review]
Thanks for the patch, @Epigene! I like the overall approach! Definitely appreciate the docs ❤️ Just a few comments, nothing major.

**Comment by ridiculous** [comment]
Those spec failures are interesting, which Ruby are you using? I can't reproduce it

File: lib/ryan.rb (hunk 2 above)
Comment (Epigene) [reply lib/ryan.rb]: Refactored.

File: lib/ryan.rb (hunk 3 above)
Comment (Epigene) [reply lib/ryan.rb]: Got rid of mode argument, performing conditional file read instead.

File: ryan.gemspec (hunk 4 above)
Comment (Epigene) [reply ryan.gemspec]: Lol, absolutely! I am so used to development deps being last in Gemfiles and gemspecs that I just dropped it in as the last one.

**Comment by Epigene** [comment]
Thanks for the quick feedback!
I pushed some changes according to your suggestions.

As for the spec failures, I am using Ruby `ruby 1.9.3p551 (2014-11-13 revision 48407) [x86_64-linux]` (as per locked spec, though there is a fair chance you originally developed this on ruby 2.0.0 with byebug)

Gem versions:

```
➜  ryan git:(initialize-from-code-string) ✗ bin/setup
Using rake 10.5.0
Using coderay 1.1.1
Using diff-lcs 1.3
Using method_source 0.8.2
Using slop 3.6.0
Using rspec-support 3.5.0
Using sexp_processor 4.8.0
Using bundler 1.14.4
Using pry 0.10.4
Using rspec-core 3.5.4
Using rspec-expectations 3.5.0
Using rspec-mocks 3.5.0
Using ruby_parser 3.8.4
Using rspec 3.5.0
Using ruby2ruby 2.3.2
Using ryan 1.1.0 from source at `.`
```

File: spec/lib/ryan.rb (hunk 5 above)
Comment (Epigene) [reply spec/lib/ryan.rb]: Any way to alias :described_class to just :klass? :D

Comment (ridiculous) [reply spec/lib/ryan.rb]: haha I guess you could do `let(:klass) { described_class }`. I'm fine though with leaving it as-is

**Comment by ridiculous** [comment]
Hmm, I haven't use 1.9.3 for years. The only Ruby I listed in .travis.yml is 2.2.1, which is probably the version I developed against. However, the gem should work on that version, and it's dependencies (Ruby2Ruby and RubyParser) work on even older versions. Fortunately it's just the spec that has issues. You can add this to spec_helper and replace usages of `Kernel.const_get` with it:
```ruby
def lookup_constant(const)
  const.to_s.split('::').inject(Object) { |k, part| k.const_get(part) }
end
```

File: lib/ryan.rb [hunk 6]
```diff
@@ -18,9 +18,19 @@ class Ryan
   def_delegators :const,
     :name, :funcs, :type, :initialization_args, :func_by_name, :class?, :module?
 
+  def self.root
+    root_path = Pathname.new File.expand_path('../..', __FILE__)
+
+    return root_path
```
Comment (ridiculous) [inline lib/ryan.rb]: Please no `return` 😆

File: lib/ryan.rb [hunk 7]
```diff
@@ -18,9 +18,19 @@ class Ryan
...
+  end
+
   # @param [Pathname, String] file
```
Comment (ridiculous) [inline lib/ryan.rb]: Let's update this comment to match the new var name of `input`

**Review by ridiculous — APPROVED** [review]

File: lib/ryan.rb (hunk 6 above)
Comment (Epigene) [reply lib/ryan.rb]: I prefer to end all methods with a return line. Being explicit about what the method returns saves headache down the road.

**Comment by ridiculous** [comment]
Looks great! There is some stuff I'll cleanup in a followup PR, such as renaming and fixing the ryan.rb spec, but that's minor. Thanks so much! I'll release version 1.1.0 later today

**Comment by Epigene** [comment]
Excellent!

**Pull request merged**